### PR TITLE
Corrupt image files are accepted, processable_image validation added : Close #9863

### LINKED ISSUE #9863: Corrupt image files are accepted and raise errors later
## Description

A user can upload files which are named like example.jpg but are not valid images. A [new version of active_storage_validations](https://github.com/openfoodfoundation/openfoodnetwork/pull/9849) can now check this, too.

## Expected Behavior

Uploading a non-image file as image should display an error straight away.

## Actual Behaviour

The upload succeeds but the image can't be displayed

## Steps to Reproduce

1. Create a text file.
2. Change the filename ending to `.jpg`.
3. Upload the .jpg file as product image.
4. Visit the shopfront to see the product image.

## Workaround

Upload a valid image.

## Severity

bug-s4: it's annoying, but you can use it

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

## Possible Fix

Add `processable_image: true` to all image validations as described in https://github.com/igorkasyanchuk/active_storage_validations/pull/168/files.

### PR DESCRIPTION
**Pull request by zafar-hussain**
#### What? Why?

Close #9863

A user can upload files which are named like example.jpg but are not valid images, upload succeeds but the image can't be displayed.

#### What should we test?
Upload an invalid image, validation will generate an error

 

#### Release notes

Changelog Category: User facing changes 

processable_image validation added

#### Dependencies
None

#### Documentation updates
One would needs to look at the error message that should be generated, in case an invalid image is used.

### CODE REVIEW

**Comment by sigmundpetersen** [comment]
I guess this is blocked by #9849 being merged first.

**Comment by jibees** [comment]
> I guess this is blocked by #9849 being merged first.

#9849 has been merged in the meantime, so I guess this one needs to be rebased onto master

File: app/models/enterprise.rb [hunk 1]
```diff
@@ -97,8 +97,14 @@ class Enterprise < ApplicationRecord
   has_one_attached :promo_image
   has_one_attached :terms_and_conditions
 
-  validates :logo, content_type: %r{\Aimage/(png|jpeg|gif|jpg|svg\+xml|webp)\Z}
-  validates :promo_image, content_type: %r{\Aimage/(png|jpeg|gif|jpg|svg\+xml|webp)\Z}
+  validates :logo,
+             processable_image: true, # added processable_image: true
```
Comment (mkllnk) [inline app/models/enterprise.rb:101]: The comment is just repeating what we can see already. Please remove.

File: app/models/enterprise.rb [hunk 2]
```diff
@@ -97,8 +97,14 @@ class Enterprise < ApplicationRecord
...
+             content_type: %r{\Aimage/(png|jpeg|gif|jpg|svg\+xml|webp)\Z}
```
Comment (mkllnk) [inline app/models/enterprise.rb:102]: You introduced some additional whitespace which should be removed. You can run rubocop locally to make sure that everything passes.

File: app/models/enterprise.rb [hunk 3]
```diff
@@ -97,8 +97,14 @@ class Enterprise < ApplicationRecord
...
   has_one_attached :terms_and_conditions
 
-  validates :logo, content_type: %r{\Aimage/(png|jpeg|gif|jpg|svg\+xml|webp)\Z}
-  validates :promo_image, content_type: %r{\Aimage/(png|jpeg|gif|jpg|svg\+xml|webp)\Z}
+  validates :logo,
+             processable_image: true, # added processable_image: true
+             content_type: %r{\Aimage/(png|jpeg|gif|jpg|svg\+xml|webp)\Z}  
+
+  validates :promo_image, 
+              processable_image: true, # added processable_image: true
```
Comment (mkllnk) [inline app/models/enterprise.rb:105]: The indentation doesn't look right here.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Great. Your code changes should do exactly what we need. We do have some style guidelines though which would be good to comply with. It makes it easier for everyone to read the code.

File: app/models/enterprise.rb (hunk 2 above)
Comment (zafar-hussain) [reply app/models/enterprise.rb:102]: removed whitespaces

File: app/models/enterprise.rb (hunk 3 above)
Comment (zafar-hussain) [reply app/models/enterprise.rb:105]: indentation made confirmant to the style

File: app/models/enterprise.rb (hunk 1 above)
Comment (zafar-hussain) [reply app/models/enterprise.rb:101]: comment removed

**Comment by jibees** [comment]
Hi @zafar-hussain 
Did you forgot to push your branch? I don't see any modification to be compliant with rubocop (warnings are still here). 
Also, did you rebase your branch onto master to take into account then merge of the PR https://github.com/openfoodfoundation/openfoodnetwork/pull/9849 ?

**Comment by zafar-hussain** [comment]
Hi @jibees,

I will push it momentarily, just sorting out the indentations

cheers

**Comment by mkllnk** [comment]
Thank you. The code changes look good to me but we have many failing specs. I'm wondering if that's a bug in the active-record-validations gem. Do you have time to look into this? If you can't find any error in our code then maybe look for open issues on that repository and open an issue if none doesn't exist yet.

**Comment by zafar-hussain** [comment]
Hi @mkllnk,

Any chance this PR can be merged before the end of October'22, i.e. by the end of Hacktoberfest

Thanks for your help,

cheers

**Comment by sigmundpetersen** [comment]
Hi @zafar-hussain 

The PR needs to pass our test suite to be approved and merged.
As Maikel commented there are many failures, so those would need to be fixed first I'm afraid.

**Comment by sigmundpetersen** [comment]
It actually seems there was a bug in `activa_storage_validations` that may be fixed by a recent update.
So if maybe @jibees could approve and merge #9916 then you can rebase this PR and there may be a chance that this can be merged before midnight 🤞 🙂

**Comment by jibees** [comment]
It's actually merged.

**Comment by mkllnk** [comment]
I rebased and the build looks better. Sill some failures though.

**Comment by RachL** [comment]
Hello @zafar-hussain Just checking : are you still working or planning to work on this PR? Don't hesitate if you need help!

**Comment by zafar-hussain** [comment]
Hi @RachL,


I am afraid, won’t be able to work on this PR, hence I am closing it

cheers

**Pull request closed without merging**